Sources/Gate/TypeTuple.swift

```diff
@@ -1,31 +1,9 @@
 struct TypeTuple: Hashable {
-    let types: String
+    let left: ObjectIdentifier
+    let right: ObjectIdentifier
     
-    init<User,Object>(_ left: User.Type, _ right: Object.Type) {
-        types = "\(left)\(right)"
-    }
-}
-
-extension Dictionary where Key == TypeTuple, Value == [Any] {
-    subscript<User,Object,Ability>(_ userType: User.Type, _ objectType: Object.Type, _: Ability.Type) -> [Policy<User,Object,Ability>] {
-        get {
-            return self[userType, objectType]
-        }
-        
-        set(newValue) {
-            self[userType, objectType] = newValue
-        }
-    }
-    
-    subscript<User,Object,Ability>(_ userType: User.Type, _ objectType: Object.Type) -> [Policy<User,Object,Ability>] {
-        get {
-            let hash = TypeTuple(User.self, Object.self)
-            return self[hash] as? [Policy<User,Object,Ability>] ?? []
-        }
-        
-        set(newValue) {
-            let hash = TypeTuple(User.self, Object.self)
-            self[hash] = newValue
-        }
+    init(_ left: Any.Type, _ right: Any.Type) {
```

Comment:
While this is technically correct and compiles you would get more performance and technical correctness if you make the two `Any.Type`'s generic. This signature isn't asking for any metatype, but quite literally the metatype associated with an implementation of `Any`.

```swift
init<Left, Right>(_ left: Left.Type, _ right: Right.Type)
```

Other than that I think it's a big improvement in simplicity. Less code quite often does mean more readability and performance 👍

Reply:
You can also make the `TypeTuple` equatable. This would only require adding an `==` function that decides the two `TypeTuples` (lhs and rhs) are equal. So you could just implement it with `lhs.left == rhs.left && lhs.right == rhs.right`.

Reply:
It's already equatable isn't it? The compiler synthesised the code necessary to make it `Hashable`, which also makes it immediately `Equatable`.